src/Mod/TechDraw/Gui/ViewMakers.cpp

```diff
@@ -0,0 +1,491 @@
+//       makexxx can be used on the App side without the Gui.
+
+//! ViewMakers is a collection of methods for creating views on a page and is
+//! phase 1 of a refactoring effort to reduce duplicated code in commands and task
+//! dialogs.
+
+
+#include "PreCompiled.h"
+#ifndef _PreComp_
+#endif
+
+#include <QMessageBox>
+
+#include <App/Document.h>
+#include <App/DocumentObject.h>
+#include <Base/Tools.h>
+#include <Gui/Command.h>
+#include <Gui/MainWindow.h>
+#include <Gui/Selection/Selection.h>
+#include <Gui/Selection/SelectionObject.h>
+
+#include <Mod/Spreadsheet/App/Sheet.h>
+
+#include <Mod/TechDraw/App/DrawPage.h>
+#include <Mod/TechDraw/App/DrawView.h>
+#include <Mod/TechDraw/App/DrawViewPart.h>
+#include <Mod/TechDraw/App/DrawViewSpreadsheet.h>
+#include <Mod/TechDraw/App/DrawViewClip.h>
+#include <Mod/TechDraw/App/DrawViewDraft.h>
+#include <Mod/TechDraw/App/DrawViewArch.h>
+#include <Mod/TechDraw/App/DrawViewImage.h>
+#include <Mod/TechDraw/App/DrawBrokenView.h>
+#include <Mod/TechDraw/App/DrawProjGroup.h>
+#include <Mod/TechDraw/App/DrawUtil.h>
+#include <Mod/TechDraw/App/Preferences.h>
+
+#include "PreferencesGui.h"
+#include "DrawGuiUtil.h"
+#include "CommandHelpers.h"
+#include "ViewMakers.h"
+
+using namespace TechDraw;
+using namespace TechDrawGui;
+using DU = DrawUtil;
+
+// TODO: wf: need to establish where selection validation is done and
+//       where the projection direction is set.
+
+//! returns a DrawViewPart based on the current selection.
+DrawViewPart* ViewMakers::makeShapeViewSelection(DrawPage* page)
+{
+    std::vector<App::DocumentObject*> shapeObjs;
+    std::vector<App::DocumentObject*> xShapeObjs;
+    CommandHelpers::getShapeObjectsFromSelection(shapeObjs, xShapeObjs);
+    return makeShapeView(page, shapeObjs, xShapeObjs);
+}
+
+
+//! returns a DrawViewPart based on the passed objects and links.
+DrawViewPart* ViewMakers::makeShapeView(DrawPage* page,
+                                        const std::vector<App::DocumentObject *> &shapeObjs,
+                                        const std::vector<App::DocumentObject *> &xShapeObjs)
+{
+    std::string FeatName = page->getDocument()->getUniqueObjectName("View");
+    std::string PageName = page->getNameInDocument();
+
+    //NOLINTBEGIN
+    Gui::Command::doCommand(Gui::Command::Doc, "App.activeDocument().addObject('TechDraw::DrawProjGroupItem', '%s')",
+                   FeatName.c_str());
+    Gui::Command::doCommand(Gui::Command::Doc, "App.activeDocument().%s.translateLabel('DrawProjGroupItem', 'View', '%s')",
+                   FeatName.c_str(), FeatName.c_str());
+    Gui::Command::doCommand(Gui::Command::Doc, "App.activeDocument().%s.addView(App.activeDocument().%s)", PageName.c_str(),
+                   FeatName.c_str());
+    //NOLINTEND
+
+    App::DocumentObject* docObj = page->getDocument()->getObject(FeatName.c_str());
+    auto* dvp = dynamic_cast<TechDraw::DrawViewPart*>(docObj);
```

Comment:
freecad_dynamic_cast

Reply:
what does freecad_dynamic_cast do?

Reply:
`dynamic_cast` uses virtual table, which is super slow
`freecad_dynamic_cast` uses member variable to check type. Is equivalent to `isDerivedFrom` +  `static_cast`